Separate services (v2)

belak opened the pull request.

jsvana committed:
Spec out separate chat backends

This is super rough. The idea is that each service will spin up a gRPC
server and a connection to Core. Core will be able to dispatch client
events to the chat backend, and the chat backend will be able to
interact with Core.

jsvana committed:
Add example relay plugin and sprinkle in some fake awaits

jsvana committed:
Forgot to add name

belak committed:
Update separate services changes based on discussions

belak committed:
Add note about help command

belak committed:
Rename StreamEvents to IngestEvents for the ChatIngest service

belak committed:
Minor renames

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,196 @@ option go_package = ".;pb";
-  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
-  rpc JoinChannel(JoinChannelRequest) returns (JoinChannelResponse);
-  rpc LeaveChannel(LeaveChannelRequest) returns (LeaveChannelResponse);
+message Service {
+  string id = 1;
+}
```
I'm assuming this is added to match the other items here, but is it worth having a message wrap a single ID? Do you envision a service ID having more fields than this eventually?

belak replied:
The idea behind this was to match the format of the Channel/ChannelInfo for when we add more introspection around services. I agree it's a little weird, but it seemed worthwhile for consistency.

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,196 @@ option go_package = ".;pb";
+message Channel {
+  string service_id = 1;
+  string channel_id = 2;
 }
```
I think this is cleaner, yeah. I originally used IDs for channels to allow for multiple instances of a channel with the same name on different chat services without the need for passing around a service ID too, but this seems cleaner overall.

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,196 @@ option go_package = ".;pb";
-  string full_help = 3;
+message ChannelInfo {
+  Channel channel = 1;
+  string name = 2;
```
Given that we're already disambiguating with the service ID above, and given what you're saying about chat services managing IDs, why have both `name` and `id`? I think that'll create a bit of confusion and annoyance when writing and maintaining both chat services and plugins.

belak replied:
As mentioned in Discord, this is in there for that platform. https://discord.com/developers/docs/resources/channel

Each channel has a unique ID and a public name. This struct could also add things like Topic and other metadata.

belak replied:
This has been renamed to display_name for clarity and a comment has been added about IDs.

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,196 @@ option go_package = ".;pb";
-  string reply_to = 1;
-  string sender = 2;
-  string message = 3;
+  ChannelSource source = 1;
```
I like this!

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,196 @@ option go_package = ".;pb";
 
-  // Seabird introspection
-  rpc ListStreams(ListStreamsRequest) returns (ListStreamsResponse);
-  rpc GetStreamInfo(StreamInfoRequest) returns (StreamInfoResponse);
-  rpc GetCoreInfo(CoreInfoRequest) returns (CoreInfoResponse);
+message User {
+  string service_id = 1;
+  string user_id = 2;
 }
 
-message StreamEventsRequest {
-  Identity identity = 1;
+message UserInfo {
+  User user = 1;
+  string name = 2;
+}
 
-  map<string, CommandMetadata> commands = 2;
+message Channel {
+  string service_id = 1;
+  string channel_id = 2;
 }
 
-message CommandMetadata {
-  string name = 1;
-  string short_help = 2;
-  string full_help = 3;
+message ChannelInfo {
+  Channel channel = 1;
+  string name = 2;
+  string topic = 3;
 }
```
Each of these names sound a bit backwards to me. A name like `User` suggests that it's a structure containing all relevant information about a user, and is basically what you're calling `UserInfo`. Instead, I'd do this:

```protobuf
message UserSelector {
  string service_id = 1;
  string user_id = 2;
}

message User {
  UserSelector selector = 1;
  string name = 2;
}
```

This makes it clearer (to me) when to expect or pass selection information like IDs versus expecting or passing full structures.

belak replied:
I like that - I'll make those changes for all the relevant structs.

belak replied:
These were all changed and cleaned up.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,164 @@
+  string nonce = 1;
+}
+
+// PongChatRequest will be sent in response to a PingChatEvent.
+message PongChatRequest {
+  string nonce = 1;
+}
```
I think this should be an RPC, and not a message stream.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,164 @@
+
+// When a chat service receives a PingChatRequest, it MUST respond with a
+// PongChatEvent containing the copied nonce.
+message PingChatRequest {
```
What's the use case of this? If you're heartbeating IRC servers and the like, IMO the chat services themselves should handle that. If this is to heartbeat the actual chat service from core, isn't that already done via the stream?

NickLanam replied:
Also, is it reasonable to assume that every chat service can even do this (a heartbeat that is more than `isItDown`)?

belak replied:
Doing this allows us to measure latency between core and the process running chat backends... and if those chat backends wanted to pass the same ping request on to their respective chat service, they could. However, yes, there is a heartbeat already on all gRPC server streams, so this may be a bit redundant - I can remove it for the MVP.

jsvana replied:
We can also instrument the RPC layer itself to get latency info. I’d see pings filling a different gap than that even, and I’m not sure what that gap is

belak replied:
Ping/pong events were dropped for now so I'm resolving this thread.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,164 @@
+// the first outgoing event with this ID will be used as the reply) and an inner
+// message type. Having an ID allows us to route an event back to the requestor
+// without having the chat backend implement a service.
+message ChatRequest {
+  string id = 1;
+
+  oneof inner {
+    PingChatRequest ping = 2;
+    PongChatRequest pong = 3;
+    SendMessageChatRequest send_message = 4;
+    SendPrivateMessageChatRequest send_private_message = 5;
+    JoinChannelChatRequest join_channel = 6;
+    LeaveChannelChatRequest leave_channel = 7;
+    SetChannelTopicChatRequest set_channel_topic = 8;
+  }
+}
```
This leaves out query requests like `GetChannelTopic` and the like. Should that be addressed, or are we punting? This would for example break the existing Minecraft plugin as it can't get the current topic.

belak replied:
There is no query for GetChannelTopic because the chat backend is responsible for sending a ChannelTopicChatEvent.

jsvana replied:
Is that expectation laid out somewhere?

belak replied:
I don't think so, but it should be. The contract for a chat backend should be laid out somewhere

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,164 @@
+  common.Identity identity = 1;
+}
+
+// PingChatEvent is an optional event which may be sent by the chat service to
+// test latency to seabird-core. When a ping event is sent, seabird-core will
+// respond with a PongChatRequest. The nonce sent with the ping will be returned
+// in the pong.
+message PingChatEvent {
+  string nonce = 1;
+}
+
+// PongChatEvent is a required event which must be sent by the chat service
+// after receiving a PingChatRequest from seabird-core. This is used by
+// seabird-core to test latency to chat services.
+//
+// Note that if a chat backend connects to a service which implements a
+// PING/PONG system similar to seabird, it may be worthwhile to test the full
+// round trip or at least check the latency there as well.
+message PongChatEvent {
+  string nonce = 1;
+}
```
Why make these into a stream of messages? If you're testing latency with these, the way this is being proposed tests both the latency of processing all messages, the connection latency, the overall RPC queue latency, and other stuff. I'd instead make this raw RPCs to core to eliminate the "hey I'm also working through other chat messages" latency.

Also, minor nit: the way this is named it seems like you're pinging the chat, when in fact it's the other way around. I understand that that's an effect of the `___ChatEvent` naming scheme, but I think it'd still be better to name this more accurately (and move it out of a message queue).

belak replied:
This was mostly copied from how IRC handles PING/PONG as I'm most familiar with that. The idea is that it's a full round trip - in theory, it's a baseline of how long it can take for a backend to respond to a message. For now I think I'll drop this as it's been shown that it's not as valuable as I'd thought, but I'd be curious to see what a better implementation would look like without having the chat backend implement an RPC service.

belak replied:
Ping/pong events were dropped for now so I'm resolving this thread.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,164 @@
+  }
+}
+
+// This service is exposed separately to the chat frontend. Its purpose is to
+// allow multiple different chat backends to register to Core and allow plugins
+// to communicate with them.
```
Once this is written, these comments need to be fleshed out and added to a set of docs in the proto repo, similar to what we did for the frontend. I can take that up once we land this.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,164 @@
+// allow multiple different chat backends to register to Core and allow plugins
+// to communicate with them.
+service ChatIngest {
+  rpc IngestEvents(stream ChatRequest) returns (stream ChatEvent);
```
I understand why you're using a bidir stream here, but I think you're forcing too much of the "chat to core" logic into it. What's the benefit of doing this?

belak replied:
I'm not aware of another alternative - doing it as a bidir stream allows us to remove the requirement that the chat backend exports a separate gRPC service.

belak replied:
Ping/pong events were dropped for now so I'm resolving this thread.

jsvana reviewed (changes requested):
Overall, this seems much better! I have some architectural-style comments inline. Tried to avoid a bunch of nits :)

NickLanam commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,164 @@
+// event is sent, the connection will be closed.
+message HelloChatEvent {
+  common.Identity identity = 1;
+}
```
Is this the only part of the handshake? Does it have both a request and a response associated with it? What about reconnects? I don't see another event type that answers these questions, but I may be overlooking something.

belak replied:
Currently, yes. This message is here so the identity token can be sent and so when we want to add more information to it (like all the channels we're currently in, stuff like that) there's a good place to put it.

belak committed:
Address some PR feedback

belak committed:
Address some more PR feedback

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,150 @@
+// ChatEvent, or start with Chat to prevent namespace conflicts between the main
+// seabird.proto.
+
+message ChatUser {
+  string id = 1;
+  string display_name = 2;
+}
+
+message ChatChannel {
+  string id = 1;
+  string display_name = 2;
+}
```
Why not use the same messages as you have in `seabird.proto`? `Channel` has one additional field, but it seems weird to have these be separate.

belak replied:
I have them separate because they have an additional field. That's exactly the reason. I'd rather not return a partially filled out message when possible.

jsvana replied:
I see. Why doesn't `ChatChannel` have the `topic` field? And `ChatUser` and `User` have the same fields. Why shouldn't they have that extra data in chat backends?

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,150 @@
+  string reason = 1;
+}
+
+message MessageChatEvent {
+  ChatUser user = 1;
+  ChatChannel channel = 2;
+  string message = 3;
+}
+
+message PrivateMessageChatEvent {
+  ChatUser user = 1;
+  ChatChannel channel = 2;
+  string message = 3;
+}
+
+message MentionChatEvent {
+  ChatUser user = 1;
+  ChatChannel channel = 2;
+  string message = 3;
+}
+
+message CommandChatEvent {
+  ChatUser user = 1;
+  ChatChannel channel = 2;
+  string command = 3;
+  string arg = 4;
+}
+
+message ChannelTopicChatEvent {
+  ChatChannel channel = 1;
+  string topic = 2;
+}
+
+message ChannelJoinEvent {
+  ChatChannel channel = 1;
+}
+
+message ChannelLeaveEvent {
+  ChatChannel channel = 1;
+}
```
How are you going to tie these back to the messages that you send? Should core be generating a message ID that it can then get back from chat backends? Otherwise it seems impossible to link events properly.

Also, how are you planning to track receipt of response from messages that core sends to chat backends?

belak replied:
The container message has an ID included. It's included with any request and event... any event spawned as a result of a request should include that ID. We also have generic success/failed events which can be used as markers.

jsvana replied:
Ah, totally missed that. Sorry!

jsvana reviewed (commented):
Seems just about there. Just have questions around a couple of things.

belak committed:
Large cleanup

- Fix README documentation
- Improve consistency in verbage (chat backend, plugin)
- Move common events to common.proto
- Add basic documentation for all events and requests

jsvana commented on README.md:
```diff
@@ -14,26 +14,23 @@ This project moves Seabird plugins from the same process as Seabird proper out i
 3. This setup allows for testing plugin instances against the production version of the `core`.
 
-## Isn't this a bad idea? I mean, this is just an IRC bot.
+## Isn't this a bad idea? I mean, this is just a chat bot.
```
:heart:

jsvana commented on README.md:
```diff
@@ -53,7 +50,7 @@ Yes. Additionally, there is no limitation on opened event streams by clients.
 Unfortunately, this is under active development and still fairly clunky, but if you really want to get involved you're welcome to!
 
-You'll need at a minimum a running instance of [core](https://github.com/seabird-irc/seabird-core). From there, you may clone various plugin repositories as desired and configure them to connect to `core`.
+At a minimum, you'll need a running instance of [core](https://github.com/seabird-irc/seabird-core). From there, you may clone various plugin repositories as desired and configure them to connect to `core`.
```
This should include a note about having at least one chat backend

jsvana commented on common.proto:
```diff
@@ -0,0 +1,78 @@
+package common;
+
+// Identity is included in every message that requires auth. One AuthMethod at a
+// time is supported. For now, the only supported method is a token. These need
```
```
// time is supported. For now, the only supported method is a token. These tokens need
```

jsvana commented on common.proto:
```diff
@@ -0,0 +1,78 @@
+  }
+}
+
+// When sent by a chat backend, the ID MAY optionally be used as a hint to the
+// frontend to get the same external ID on reconnection. When sent by core to a
```
What's "frontend"? What happens if the ID isn't used?

belak replied:
By frontend, I meant core.

jsvana replied:
Can you update the comment? Also what about the second question?

jsvana replied:
Still need to change this to `core`

jsvana commented on common.proto:
```diff
@@ -0,0 +1,78 @@
+  string type = 2;
+}
+
+// When sent by a chat backend, the ID MUST be a unique ID representing this
```
Unique to the backend? To core? What happens if the ID isn't unique? Errors?

belak replied:
Unique to core. The connection will be closed if there is already a running backend of the same type with the same ID.

jsvana replied:
Can you write both of those explicitly here?

jsvana replied:
Also, if things are unique to core, shouldn't core expose a "generate me a unique ID" RPC?

jsvana replied:
Chatted offline. The IDs are meant to be a descriptive string and not a random string (what I thought you were proposing). I think calling both the structure and the behaviors on collision out explicitly in the comment is important.

jsvana commented on common.proto:
```diff
@@ -0,0 +1,78 @@
+  string display_name = 2;
+}
+
+// When sent by a chat backend, the ID MUST be a unique ID representing this
```
Same question as above.

jsvana commented on common.proto:
```diff
@@ -0,0 +1,78 @@
+// rather than just an ID so any clients can have access to the name as well.
+// This may change in the future.
+message ChannelSource {
+  User user = 1;
+  string channel_id = 2;
```
nit: in my head these should be reversed because you're saying this is from a channel, and that's the second field lol

jsvana commented on common.proto:
```diff
@@ -0,0 +1,78 @@
+  string text = 2;
+}
+
+// MentionEvent will be sent when a user mentions the chat backend at the start
```
`...mentions the _bot_ at the start`, no?

jsvana commented on common.proto:
```diff
@@ -0,0 +1,78 @@
+  string text = 2;
+}
+
+// CommandEvent will be sent when a user issues a command in a channel. Commands
+// cannot be issued via private message.
```
Is this true of the current iteration of Seabird? Interesting regression I didn't think about.

belak replied:
Yep, this is true for the current implementation

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,163 @@ option go_package = ".;pb";
-  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
-  rpc JoinChannel(JoinChannelRequest) returns (JoinChannelResponse);
-  rpc LeaveChannel(LeaveChannelRequest) returns (LeaveChannelResponse);
+// Note that the IDs exposed by the Service, User, and Channel are compound IDs
```
Details on what `compound` means? Maybe an example? I would either say "they're compound and here's an example" or say "do not depend on the structure of these IDs", otherwise plugins might depend on the structure without it being guaranteed to stay the same.

jsvana replied:
Also which service will route these? If you're talking about routing to the backend, does that mean `ChannelSource` is actually `ChannelDestination`? I don't think you're talking about plugin routing as that's not really supported.

belak replied:
I changed "compound" to "opaque" and clarified that core will route relevant items to the proper backend

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,163 @@ option go_package = ".;pb";
+  common.Identity identity = 1;
 
+  // A registry of commands this plugin responds to. A plugin MUST register all
+  // commands it responds to in order to receive those events. The
```
`...order to receive those events as commands instead of MessageEvents`, right?

belak replied:
No, if a message is parsed as a command, it will not be another message type.

jsvana replied:
I'd add that here as well then.

jsvana replied:
Please add a note similar to what you added in the command message comment

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,163 @@ option go_package = ".;pb";
 
+  // A registry of commands this plugin responds to. A plugin MUST register all
+  // commands it responds to in order to receive those events. The
+  // CommandMetadata's name MUST match the map key.
```
Is this intended to return an error?

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,163 @@ option go_package = ".;pb";
   map<string, CommandMetadata> commands = 2;
 }
 
+// CommandMetadata groups together a command's name along with short in-line
+// help and the full private help.
```
Maybe include a note about how short vs long might look?

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,163 @@ option go_package = ".;pb";
 
-message User {
-  string nick = 1;
+  // NOTE: this channel_name is the only place name is used for an identifier -
```
Add `all other times channels will be referred to by ID`

jsvana commented on seabird.proto:
```diff
@@ -4,168 +4,163 @@ option go_package = ".;pb";
 
-  string name = 2;
-  string message = 3;
+message BackendInfoResponse {
+  common.Backend service = 1;
 }
```
Should this be a shallow list of channel IDs too or is that too bulky?

belak replied:
I'm going to punt on this because it can be added later.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+
+package seabird;
+
+// Note that all of the types in this file end with ChatRequest or end with
+// ChatEvent to prevent namespace conflicts between the main seabird.proto file.
```
Why is this necessary? Protobuf imports are already namespaced, and you control naming on code import.

belak replied:
With go when you build all these protos into the same go package, they can conflict.

jsvana replied:
That begs the question though: why are these the same Go package?

belak replied:
Common practice is to put generated protobufs in a `pb` package, even when coming from multiple protobufs. For some reason when you `import` another protobuf in Go, it imports it into the same namespace, so those common.User messages are still just exported as User.

belak replied:
Punting on this for now

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+// ChatEvent to prevent namespace conflicts between the main seabird.proto file.
+
+// SendMessageChatRequest requests for a message to be sent to a given channel.
+// Either SuccessChatEvent or FailedChatEvent SHOULD be returned in response if
+// possible. This MUST NOT cause any chat events from the bot to appear in the
```
`SHOULD be returned...if possible.` -- what happens if not, and why is it beneficial? What will core do with them?

belak replied:
Core will return a timeout error if it isn’t received in time

jsvana replied:
I'd still like to see each of the comments about failure characteristics pulled out and added above this

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+
+// SendMessageChatRequest requests for a message to be sent to a given channel.
+// Either SuccessChatEvent or FailedChatEvent SHOULD be returned in response if
+// possible. This MUST NOT cause any chat events from the bot to appear in the
+// output event stream.
```
I don't understand the third sentence of this comment:
```
// ...This MUST NOT cause any chat events from the bot to appear in the
// output event stream.
```

What does that mean?

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+}
+
+// SendPrivateMessageChatRequest requests for a message to be sent to a given
+// user. Either SuccessChatEvent or FailedChatEvent SHOULD be returned in
+// response if possible. This MUST NOT cause any chat events from the bot to
+// appear in the output event stream.
```
I'd pull all of the repeated docs for messages out and define things once.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+}
+
+// Each ChatRequest has an ID (which can be attached to outgoing events - only
+// the first outgoing event with this ID will be used as the reply) and an inner
```
This is defined from the perspective of core so should this say `incoming to core` instead of `outgoing`? What happens to other events? Can you explicitly call it out if they simply get dropped?

jsvana replied:
I would also add expectations around what happens when `Request`s don't get a `Response`.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+  }
+}
+
+// HelloChatEvent must be the first event sent by the chat backend. If a
+// different event is sent, the connection will be closed.
+message HelloChatEvent {
+  common.Identity identity = 1;
+
+  common.Backend service_type = 2;
+}
```
Does this just exist because bidir streams don't accept an argument as a message?

belak replied:
Yeah, you can't accept both a single message and a stream.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+}
+
+// HelloChatEvent must be the first event sent by the chat backend. If a
+// different event is sent, the connection will be closed.
```
Love the explicit callout!

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+message SuccessChatEvent {}
+
+// This is a marker event to respond to seabird-core that performing an action
+// has failed. Note that an action which takes too long will also be marked as
+// failed automatically. The time alotted for an action is not static and may
+// change based on load.
```
This bit should be called out above. Also, expressing that you have a deadline and then saying that the deadline changes makes this not a deadline IMO.

belak replied:
Yeah, I didn't call it a deadline for exactly that reason, I just wanted to specify that core will only wait for an unspecified amount of time before considering a request failed.

jsvana replied:
That doesn't provide the chat backend with any information, though. That's sort of the same as saying "your request could be failed immediately if there's too much load", which is fine, but don't express it as a timeout, express it as backpressure.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+syntax = "proto3";
```
Are there any events that a chat backend could send to core and expect a response for? It seems like all now are either `respond to core` or `do something with no ack`.

belak replied:
Anything where a chat backend would call into core would be implemented as rpc rather than messages on the stream.

jsvana replied:
Does that mean that LeaveChannel should be an RPC instead? What happens if the bot is kicked?

belak replied:
No, this is specifically for chat backends calling in to core, so if an event happens, it should be exposed via the stream. If the backend actually needs to do something, we will need to add APIs for it to call in to core.

Also, if a bot was kicked, a leave channel event will be generated

jsvana replied:
> Anything where a chat backend would call into core would be implemented as rpc rather than messages on the stream.
> No, this is specifically for chat backends calling in to core, so if an event happens, it should be exposed via the stream.

These two messages are saying opposite things. Which is it? The first half of your response also seems to contradict the second half :(

> Also, if a bot was kicked, a leave channel event will be generated

This goes against what you said as well--why wouldn't this be an RPC?

jsvana replied:
We resolved this offline.

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -0,0 +1,144 @@
+// here will not be mapped to UUIDs and some events are only to support the data
+// that seabird-core tracks.
+//
+// HelloChatEvent is the only message type that is required for a chat backend
+// to work.
```
Is there a threshold of failed events that will cause a disconnect?

belak replied:
Not currently - I was basing this on IRC where you might not have a confirmation that you did something.

belak replied:
Also, specifying you only need to respond to Hello means if we add more event types and backends don't support them they will continue to work.

jsvana replied:
Right, no issues with that. My only concern was calling out whether or not there's a threshold of failed events. I'd include that (even if there isn't one, call it out) in the "failure behaviors" section

belak replied:
I'd rather not explicitly define this quite yet - I'd like to revisit this when we start to add more event types because I think a full solution would include the backend specifying which types of events it supports and I don't want to figure that out just yet.

jsvana replied:
Fair, but backend types are orthogonal. At least call out that chat backends can't currently make assumptions around failure events and timeouts.

jsvana reviewed (commented):
I love all of the comments!

belak committed:
Update documentation

jsvana commented on common.proto:
```diff
@@ -21,17 +21,19 @@ message Backend {
+// When sent by a chat backend, the ID MUST be unique per users in that running
+// backend. Failure to follow this will result in incorrect information being
+// sent to plugins. When sent by core to a plugin, the ID MUST be an ID pointing
+// to an instance of a user on a service.
```
`...of a user on a chat backend.`?

jsvana commented on common.proto:
```diff
@@ -21,17 +21,19 @@ message Backend {
-// When sent by a chat backend, the ID MUST be a unique ID representing this
-// channel in the given chat service. When sent by core to a plugin, the ID MUST
-// be an ID pointing to an instance of a channel on a service.
+// When sent by a chat backend, the ID MUST be unique per channels in that
```
`...unique across channels...`

jsvana commented on common.proto:
```diff
@@ -21,17 +21,19 @@ message Backend {
-// When sent by a chat backend, the ID MUST be a unique ID representing this
-// user in the given chat service. When sent by core to a plugin, the ID MUST be
-// an ID pointing to an instance of a user on a service.
+// When sent by a chat backend, the ID MUST be unique per users in that running
```
`...unique across users...`

jsvana commented on common.proto:
```diff
@@ -21,17 +21,19 @@ message Backend {
+// When sent by a chat backend, the ID MUST be unique per channels in that
+// running backend. Failure to follow this will result in incorrect information
+// being sent to plugins. When sent by core to a plugin, the ID MUST be an ID
+// pointing to an instance of a channel on a service.
```
`...of a channel on a chat backend.`

jsvana commented on common.proto:
```diff
@@ -70,7 +72,8 @@ message MentionEvent {
 
 // CommandEvent will be sent when a user issues a command in a channel. Commands
-// cannot be issued via private message.
+// cannot be issued via private message. If a message is parsed as a command, it
+// SHOULD NOT be sent as another message type.
```
I'm not a huge fan of "SHOULD NOT". It doesn't give the client any guarantees. Instead, I'd say:

```
If a message is parsed as a command, plugins must not expect to receive the message as another type as well.
```

belak replied:
I'm actually going to change this to a MUST NOT to make the chat backend contract clearer

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -65,12 +65,19 @@ message ChatRequest {
 
-// HelloChatEvent must be the first event sent by the chat backend. If a
-// different event is sent, the connection will be closed.
+// HelloChatEvent MUST be the first event sent by the chat backend. If a
+// different event is sent, the connection will be closed. The chat backend MUST
+// provide an ID unique between all running instances of this type in the
+// backend_info. If the ID is not provided or is not unique, the connection will
+// be closed.
+//
+// Note that all IDs sent by a chat backend will be automatically namespaced by
+// core based on the backend. As an example, a backend of type "foo" and an ID
+// of "bar" may show up as something like "foo.bar" to plugins.
```
This is awesome!

jsvana commented on seabird_chat_ingest.proto:
```diff
@@ -114,6 +120,11 @@ message ChangeChannelEvent {
 // HelloChatEvent is the only message type that is required for a chat backend
 // to work.
+//
+// Core may be configured with both time and load-based request failure
```
`...both time- and...`

belak replied:
I've never seen this syntax used before... maybe `both time-based and load-based` would make sense?

belak committed:
Address PR feedback

belak committed:
More PR feedback
